### PR TITLE
Remove buf copy when the compressor exist

### PR DESCRIPTION
**Pull request by ZhouyihaiDing**
Re-use the buffer used by compressor. Doesn't need to create new "buf" and copy the old data.

### CODE REVIEW

File: rpc_util.go [hunk 1]
```diff
@@ -291,8 +291,10 @@ func (p *parser) recvMsg(maxReceiveMessageSize int) (pf payloadFormat, msg []byt
 // generates the message header of 0 message length.
 func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayload *stats.OutPayload) ([]byte, error) {
 	var (
-		b      []byte
-		length uint
+		b          []byte
+		length     uint
+		payloadLen = 1
```
Comment (dfawley) [inline rpc_util.go]: These two should be `const` instead.

File: rpc_util.go [hunk 2]
```diff
@@ -307,40 +309,45 @@ func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayl
...
 			}
 			b = cbuf.Bytes()
+			length = uint(len(b) - 5)
+		} else {
+			length = uint(len(b))
 		}
-		length = uint(len(b))
 	}
+
 	if length > math.MaxUint32 {
```
Comment (dfawley) [inline rpc_util.go]: This is unlikely to even trigger if there is a 4GB message.  length should be a uint64 if we're serious about this.

File: rpc_util.go [hunk 3]
```diff
@@ -307,40 +309,45 @@ func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayl
 			outPayload.Data = b
 			outPayload.Length = len(b)
 		}
+
 		if cp != nil {
+			// pre-alloc place
+			cbuf.Write(make([]byte, 5))
```
Comment (dfawley) [inline rpc_util.go]: Be consistent about your use of constants -- use "payloadLen+sizeLen" where you have 5.

File: rpc_util.go [hunk 4]
```diff
@@ -307,40 +309,45 @@ func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayl
...
-		payloadLen = 1
-		sizeLen    = 4
-	)
-
-	var buf = make([]byte, payloadLen+sizeLen+len(b))
+	if cp != nil {
+		// Write payload format
+		b[0] = byte(compressionMade)
+		// Write length of b into buf
+		binary.BigEndian.PutUint32(b[1:], uint32(length))
```
Comment (dfawley) [inline rpc_util.go]: 1->payloadLen

File: rpc_util.go [hunk 5]
```diff
@@ -307,40 +309,45 @@ func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayl
...
+		if outPayload != nil {
+			outPayload.WireLength = len(b)
+		}
+		return b, nil
+	}
 
+	var buf []byte
```
Comment (dfawley) [inline rpc_util.go]: How about inverting things to share the length-writing and wirelength code?

```go
var buf []byte
if cp == nil {
  buf = make([]byte, payloadLen+sizeLen+len(b))
  buf[0] = byte(compressionNone)
} else {
  buf = b
  buf[0] = byte(compressionMade)
}
binary.BigEndian.PutUint32(buf[1:], uint32(length))
if outPayload != nil {
  outPayload.WireLength = len(buf)
}
return buf, nil
```

Comment (ZhouyihaiDing) [inline rpc_util.go]: You are right. I thought buf = b will assign new alloc for buf, but it turns out it doesn't.

File: rpc_util.go (hunk 2 above)
Comment (ZhouyihaiDing) [inline rpc_util.go]: If the size can be larger than 4GB, we need to add one more byte (or 2 to 3 bits?) in the header for every message sent to save the length of the message. It's that worth?

Comment (dfawley) [inline rpc_util.go]: That's not an option -- we must follow the protocol, which is 4 bytes to encode the length.

My point is about the type of the variables.  Checking whether a uint is > MaxUint32 will only work on platforms where a uint is 64 bits.

tl;dr: please make length a uint64.

Comment (ZhouyihaiDing) [inline rpc_util.go]: Since len(b) returns a int. It seems wrong because length > math.MaxUint32 will always returns false.

**Comment by dfawley** [comment]
>
> Since len(b) returns a int. It seems wrong because length > math.MaxUint32
> will always returns false.
>
Good point.  Only on those systems where "int" is 32 bits.  Presumably
those systems can't have slices bigger than 4GB, so I suppose this is
fine.  Why not leave "length" an int (instead of a uint, with casting),
then?

**Comment by ZhouyihaiDing** [comment]
Update end2end_test.go to make it pass travis. Can you help me review it again? 
Thank you very much!

File: call.go [hunk 6]
```diff
@@ -99,17 +99,29 @@ func sendRequest(ctx context.Context, dopts dialOptions, compressor Compressor,
...
+	hdr, data, err := encode(dopts.codec, args, compressor, cbuf, outPayload)
 	if err != nil {
 		return err
 	}
 	if c.maxSendMessageSize == nil {
 		return Errorf(codes.Internal, "callInfo maxSendMessageSize field uninitialized(nil)")
 	}
-	if len(outBuf) > *c.maxSendMessageSize {
-		return Errorf(codes.ResourceExhausted, "grpc: trying to send message larger than max (%d vs. %d)", len(outBuf), *c.maxSendMessageSize)
+	if len(hdr)+len(data) > *c.maxSendMessageSize {
```
Comment (dfawley) [inline call.go]: I think we said we wouldn't include the header as the message for this check.  (Also the error message would be out of sync.)

File: rpc_util.go [hunk 7]
```diff
@@ -310,39 +311,34 @@ func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayl
...
-		outPayload.WireLength = len(buf)
+		outPayload.WireLength = payloadLen + sizeLen + len(b)
 	}
 
-	return buf, nil
+	secondStart := 16384 - payloadLen - sizeLen
+	if len(b) < secondStart {
+		secondStart = len(b)
+	}
+	bufHeader = append(bufHeader, b[:secondStart]...)
```
Comment (dfawley) [inline rpc_util.go]: Oh, I see what happened here.

Please keep outputting this the way it was before: only the header + all of the message data.

Then pass them into the transport.Write function as two slices (instead of making two separate calls), then combine within there.  This way, the knowledge that a frame is 16kb is contained within the transport, which is desirable as it's really a function of the transport not gRPC.

File: call.go (hunk 6 above)
Comment (ZhouyihaiDing) [inline call.go]: Done.

File: rpc_util.go (hunk 7 above)
Comment (ZhouyihaiDing) [inline rpc_util.go]: Done. Transport.Write will append data to the header slice.

File: transport/http2_client.go [hunk 8]
```diff
@@ -763,11 +767,19 @@ func (t *http2Client) Write(s *Stream, data []byte, opts *Options) error {
 			t.writableChan <- 0
 			continue
 		}
-		if r.Len() == 0 && t.framer.adjustNumWriters(0) == 1 {
+		if r.Len() == 0 && t.framer.adjustNumWriters(0) == 1 && isLastSlice {
```
Comment (dfawley) [inline transport/http2_client.go]: How about moving this after the new if `r.Len() == 0` that you added so you don't need to check `isLastSlice`?

File: transport/http2_server.go [hunk 9]
```diff
@@ -821,8 +821,15 @@ func (t *http2Server) WriteStatus(s *Stream, st *status.Status) error {
 
 // Write converts the data into HTTP2 data frame and sends it out. Non-nil error
 // is returns if it fails (e.g., framing error, transport error).
-func (t *http2Server) Write(s *Stream, data []byte, opts *Options) (err error) {
+func (t *http2Server) Write(s *Stream, hdr []byte, data []byte, opts *Options) (err error) {
```
Comment (dfawley) [inline transport/http2_server.go]: ....soooo much duplication :/

All of the comments in http2Client.Write apply here as well.

(Don't try to fix the duplication in this PR.)

File: transport/http2_client.go [hunk 10]
```diff
@@ -763,11 +767,19 @@ func (t *http2Client) Write(s *Stream, data []byte, opts *Options) error {
...
 			// Do a force flush iff this is last frame for the entire gRPC message
 			// and the caller is the only writer at this moment.
 			forceFlush = true
 		}
+		if r.Len() == 0 {
+			if isLastSlice && opts.Last {
+				endStream = true
+			} else if !isLastSlice && len(data) != 0 {
+				r = bytes.NewBuffer(data)
```
Comment (ZhouyihaiDing) [inline transport/http2_client.go]: Not Done. I need some time to check the overhead if using slice directly.

File: transport/http2_client.go (hunk 8 above)
Comment (ZhouyihaiDing) [inline transport/http2_client.go]: Done.

File: transport/http2_client.go [hunk 11]
```diff
@@ -678,8 +678,15 @@ func (t *http2Client) GracefulClose() error {
 // should proceed only if Write returns nil.
 // TODO(zhaoq): opts.Delay is ignored in this implementation. Support it later
 // if it improves the performance.
-func (t *http2Client) Write(s *Stream, data []byte, opts *Options) error {
-	r := bytes.NewBuffer(data)
+func (t *http2Client) Write(s *Stream, hdr []byte, data []byte, opts *Options) error {
+	secondStart := 16379 // 16384 - payloadLen - sizeLen
```
Comment (ZhouyihaiDing) [inline transport/http2_client.go]: Done.
